Remora.Discord.Extensions/Services/PermissionComputationService.cs

```diff
@@ -0,0 +1,198 @@
+using Remora.Rest.Core;
+using Remora.Results;
+
+namespace Remora.Discord.Extensions.Services;
+
+/// <inheritdoc />
+public class PermissionComputationService : IPermissionComputationService
+{
+    private readonly ContextInjectionService _contextInjection;
+    private readonly IDiscordRestGuildAPI _guildAPI;
+    private readonly IDiscordRestChannelAPI _channelAPI;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="PermissionComputationService"/> class.
+    /// </summary>
+    /// <param name="contextInjection">The context injection service.</param>
+    /// <param name="guildAPI">The guild API.</param>
+    /// <param name="channelAPI">The channel API.</param>
+    public PermissionComputationService
+    (
+        ContextInjectionService contextInjection,
+        IDiscordRestGuildAPI guildAPI,
+        IDiscordRestChannelAPI channelAPI
+    )
+    {
+        _contextInjection = contextInjection;
+        _guildAPI = guildAPI;
+        _channelAPI = channelAPI;
+    }
+
+    /// <inheritdoc/>
+    public async Task<Result<IDiscordPermissionSet>> ComputeMemberPermissions
+    (
+        Snowflake userID,
+        Snowflake guildID,
+        Snowflake? channelID = null,
+        CancellationToken ct = default
+    )
+    {
+        var getMember = await _guildAPI.GetGuildMemberAsync(guildID, userID, ct);
+        if (!getMember.IsDefined(out var member))
+        {
+            return Result<IDiscordPermissionSet>.FromError(getMember);
+        }
+
+        var getGuildRoles = await _guildAPI.GetGuildRolesAsync(guildID, ct);
+        if (!getGuildRoles.IsDefined(out var guildRoles))
+        {
+            return Result<IDiscordPermissionSet>.FromError(getGuildRoles);
+        }
+
+        var everyoneRole = guildRoles.First(r => r.ID == guildID);
+        var memberRoles = guildRoles.Where(r => member.Roles.Contains(r.ID)).ToList();
+        IReadOnlyList<IPermissionOverwrite>? overwrites = null;
+
+        // ReSharper disable once InvertIf
+        if (channelID is not null)
+        {
+            var getChannel = await _channelAPI.GetChannelAsync(channelID.Value, ct);
+            if (!getChannel.IsDefined(out var channel))
+            {
+                return Result<IDiscordPermissionSet>.FromError(getChannel);
+            }
+
+            overwrites = channel.PermissionOverwrites.OrDefault();
+        }
+
+        return Result<IDiscordPermissionSet>.FromSuccess
+        (
+            DiscordPermissionSet.ComputePermissions(userID, everyoneRole, memberRoles, overwrites)
+        );
+    }
+
+    /// <inheritdoc />
+    public Task<Result<IDiscordPermissionSet>> ComputeContextualMemberPermissions
+    (
+        Snowflake userID,
+        bool inChannel = true,
+        CancellationToken ct = default
+    )
+    {
+        if (_contextInjection.Context is null)
+        {
+            throw new InvalidOperationException("Contextual permission computation requires an available context.");
```

Comment:
re: Prior comment; documentation needs to be updated, or perhaps return an error?

Reply:
In this case, an exception was the intended action. Using the contextual methods in a non-contextual context (hoo boy, words) is programmer error, not data error.